### ISSUE TITLE
TRANSPARENT PROXY

### ISSUE THREAD

**Issue by alonbg** [F-Stack/f-stack#43]
Hi,
Very exciting project.
Will nginx [transparent proxy setsockopt](https://github.com/F-Stack/f-stack/blob/master/app/nginx-1.11.10/src/event/ngx_event_connect.c#L343) work if enabled ([NGX_HAVE_TRANSPARENT_PROXY](https://github.com/F-Stack/f-stack/blob/f854666911c7482bde4b15c2f1c11dbe46de8d13/app/nginx-1.11.10/auto/unix#L336) )?

If not, what's missing for it to work?

Thank you

**Comment by whl739** [comment]
We didn't test transparent proxy with nginx.
But as i know, `IP_TRANSPARENT` is not supported by FreeBSD stack and `IP_BINDANY` is supported.
Since the `configure` checks for these options on Linux, it couldn't detective  the real supported options on F-Stack.
Maybe i should modify `auto/unix` to support `IP_BINDANY`.

**Comment by alonbg** [comment]
@whl739 thank you. I believe you are correct, `IP_BINDANY` is the FreeBSD equiv to Linux's `IP_TRANSPARENT`. This nginx feature (i.e. `NGX_HAVE_TRANSPARENT_PROXY`) is not enabled by default in `auto/unix`. I don't think it should be as it requires additional system setup [in Linux](http://tldp.org/HOWTO/TransparentProxy-6.html) and maybe on [FreeBSD](http://ninjo.blogspot.co.il/2013/08/sslh-on-freebsd-with-transparent.html) to properly function.

**Comment by alonbg** [comment]
@whl739 , considering `configure` would be modified somehow to detect ` IP_BINDANY` option ...
Can you asses please what additional configuration, tools, functionality is missing for this to work ?

**Comment by whl739** [comment]
A simple way:
```
diff --git a/app/nginx-1.11.10/auto/unix b/app/nginx-1.11.10/auto/unix
index 2354d29..0666770 100644
--- a/app/nginx-1.11.10/auto/unix
+++ b/app/nginx-1.11.10/auto/unix
@@ -365,7 +365,7 @@ ngx_feature_incs="#include <sys/socket.h>
 ngx_feature_path=
 ngx_feature_libs=
 ngx_feature_test="setsockopt(0, IPPROTO_IP, IP_TRANSPARENT, NULL, 0)"
-. auto/feature
+#. auto/feature


 # FreeBSD bind to any address for transparent proxying
@@ -377,7 +377,8 @@ ngx_feature_incs="#include <sys/socket.h>
                   #include <netinet/in.h>"
 ngx_feature_path=
 ngx_feature_libs=
-ngx_feature_test="setsockopt(0, IPPROTO_IP, IP_BINDANY, NULL, 0)"
+#ngx_feature_test="setsockopt(0, IPPROTO_IP, IP_BINDANY, NULL, 0)"
+ngx_feature_test="int a;"
 . auto/feature
```

**Comment by alonbg** [comment]
@whl739 thank you. Additionally, for this socket option to work, the FreeBSD stack should[ support iptables](http://tldp.org/HOWTO/TransparentProxy-6.html).
I assume this is not supported yet ...

**Comment by whl739** [comment]
We have already supported `ipfw`. It is similar with `iptables` in Linux.

**Comment by tigerjibo** [comment]
Hi whl739:
   How to use the ipfw tool to support the transparent proxy for the tcp or udp protocol?

**Comment by whl739** [comment]
@tigerjibo 
Sorry, i don't know more about transparent proxy.
All i know is that as @alonbg says, [this link](http://tldp.org/HOWTO/TransparentProxy-6.html) shows how to configure  transparent proxy, it must use `iptables`, and `ipfw` in F-Stack can do the same thing.

**Comment by tigerjibo** [comment]
@whl739 :    
   Thanks for your reply as soon as possible. I will modify ipfw and ipfil as to support the transparent proxy.

**Comment by oceansw** [comment]
Hi, @tigerjibo
As mentioned you have modified the ipfw to support transparent proxy, does it work now? 
Thanks in advance.

**Comment by tigerjibo** [comment]
Hi:
  For now , the code has finished ,But I don't have enough time to test it






* Best RegardsJi  Bo
(冀博)---------------------------------------------------------------------------*

2017-12-28 11:13 GMT+08:00 oceansw <notifications@github.com>:

**Comment by oceansw** [comment]
Hi,  @tigerjibo @whl739 
Many thanks for the reply. Could you share the code if possible? I can do more test .
I'm facing the same issue:
At my tproxy host (addr:IPA), by using ipfw :./ipfw/ipfw add 5000 fwd 127.0.0.1,80 tcp from any to any 80,
redirect all 80(to any dst) to IPA:127.0.0.1:80,
and int the test code :
    struct sockaddr_in my_addr;
    bzero(&my_addr, sizeof(my_addr));
    my_addr.sin_family = AF_INET;
    my_addr.sin_port = htons(80);
    my_addr.sin_addr.s_addr = htonl(INADDR_ANY);
   ff_setsockopt(sockfd, SOL_IP, IP_BINDANY, &tproxy, sizeof(tproxy),

I got the connection from client to the IPA(1.1.1.1 is a HTTP server), but after TCP handshaking ,
the connection closed by IPA, pls reference attached info:
-------------TCPDUMP-------------
21:40:16.816255 IP 192.168.29.244.61007 > 1.1.1.1.http: Flags [S], seq 2344767005, win 29200, options [mss 1460,sackOK,TS val 1399758137 ecr 0,nop,wscale 7], length 0
21:40:16.816873 IP 1.1.1.1.http > 192.168.29.244.61007: Flags [S.], seq 189254640, ack 2344767006, win 8192, options [mss 1460,nop,wscale 9,sackOK,TS val 2625440235 ecr 1399758137], length 0
21:40:16.817089 IP 192.168.29.244.61007 > 1.1.1.1.http: Flags [.], ack 1, win 229, options [nop,nop,TS val 1399758138 ecr 2625440235], length 0
-------------TCPDEBUG-------------
0x7ffff7e9c820 LISTEN:input [8bc24e1d..8bc24e59)@0, urp=0<SYN> -> LISTEN
	rcv_(nxt,wnd,up) (0,0,0) snd_(una,nxt,max) (0,0,0)
	snd_(wl1,wl2,wnd) (0,0,0)
0x7ffff7e9c410 CLOSED:user ATTACH -> CLOSED
	rcv_(nxt,wnd,up) (0,0,0) snd_(una,nxt,max) (0,0,0)
	snd_(wl1,wl2,wnd) (0,0,0)
0x7ffff7e9c410 CLOSED:user ABORT -> CLOSED
	rcv_(nxt,wnd,up) (0,0,0) snd_(una,nxt,max) (0,0,0)
	snd_(wl1,wl2,wnd) (0,0,0)
0x7ffff7e9c820 CLOSED:output [b47cbf1..b47cc05)@0, urp=0<RST> -> LISTEN
	rcv_(nxt,wnd,up) (0,0,0) snd_(una,nxt,max) (0,0,0)
	snd_(wl1,wl2,wnd) (0,0,0)

Thanks in advance.

**Comment by rolfliu** [comment]
Hi @tigerjibo  I want to use nginx transparent proxy, but while test, I got a setsockopt(IP_BINDANY) error using the f-stack, could you send me the patch you modified?
Thank you very much.
My email address is rolfxy.liu@gmail.com


Best Regards!
